src/librustc_typeck/mem_categorization.rs

```diff
@@ -561,9 +581,13 @@ impl<'a, 'tcx> MemCategorizationContext<'a, 'tcx> {
             }
         };
         let mut projections = base_place.place.projections;
-        projections.push(Projection { kind: ProjectionKind::Deref(base_ty) });
+        projections.push(Projection {
+            kind: ProjectionKind::Deref,
+            before_ty: base_ty,
+            after_ty: deref_ty,
+        });
 
-        let ret = PlaceWithHirId::new(node.hir_id(), deref_ty, base_place.place.base, projections);
+        let ret = PlaceWithHirId::new(node.hir_id(), base_ty, base_place.place.base, projections);
         debug!("cat_deref ret {:?}", ret);
```

Comment:
this should be `base_place.place.base_ty`

Reply:
Done. Good catch